### ISSUE TITLE
No Brew commands work

### ISSUE THREAD

**Issue by tmongero** [Homebrew/homebrew#48256]
Hi there - seemingly out of nowhere, none of my Homebrew commands are working. I tried brew update, brew cleanup, brew doctor, even just 'brew' (Expecting the help file).. All I get is the following:

![image](https://cloud.githubusercontent.com/assets/16004728/12430122/83f15fe0-beb3-11e5-9963-66db7daca649.png)

I update homebrew every day, sometimes every couple days (I usually never go more than 3-5 days without an update). I just upgraded from 10.11.2 to 10.11.3 in hopes of some magic, but no magic happened.

Any thoughts?

**Comment by UniqMartin** [comment]
Are you able to run `brew --prefix` successfully and what does it return? Can you also please run `cd /usr/local && git status --untracked-files=all` and paste its output here? Thanks!

**Comment by tmongero** [comment]
Brew --prefix returns the same error as what is in my original screenshot of output.

![image](https://cloud.githubusercontent.com/assets/16004728/12430487/78540f14-beb5-11e5-9600-54cf5a9ea39e.png)

![image](https://cloud.githubusercontent.com/assets/16004728/12430510/9b455ae6-beb5-11e5-9492-f5ad0847bfc9.png)

![image](https://cloud.githubusercontent.com/assets/16004728/12430519/a7a60ae2-beb5-11e5-8204-e7f720a0b5f8.png)


Those are screenshots of the full output.

**Comment by UniqMartin** [comment]
This looks like something went wrong during a previous update via `brew update`. Do you remember that it printed some error the last time you used it?

If you are sure you haven't made any customizations to your Homebrew installation, please try

```
cd /usr/local && git reset --hard origin/master
```

to reset it to a working state. If that completes without error, please paste the output of `cd /usr/local && git status --untracked-files=all` again. And try `brew --prefix`, `brew config`, and `brew doctor` again.

**Comment by jeroenj** [comment]
I am seeing similar issues. Unlike @tmongero I have a clean `/usr/local` directory.

I've tracked it down to [this commit](https://github.com/Homebrew/homebrew/commit/390e327f55126f7ee9519bc6c03a0c0aa14f7168#diff-d9897ba351f5b5528d4d53b545e519acR34) which got merged two days ago. The `ENV["HOMEBREW_PREFIX"]` variable is `nil` causing `Pathname` to raise that `TypeError`.

Right now the only way to reproduce this is by using the [homebrew chef cookbook](https://github.com/chef-cookbooks/homebrew) or by trying out the following (what the cookbook basically does):

```
irb(main):001:0> $LOAD_PATH.unshift ::File.join(`brew --prefix`.chomp, 'Library', 'Homebrew')
=> ["/usr/local/Library/Homebrew", "/Library/Ruby/Site/2.0.0", "/Library/Ruby/Site/2.0.0/x86_64-darwin15", "/Library/Ruby/Site/2.0.0/universal-darwin15", "/Library/Ruby/Site", "/System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/lib/ruby/vendor_ruby/2.0.0", "/System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/lib/ruby/vendor_ruby/2.0.0/x86_64-darwin15", "/System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/lib/ruby/vendor_ruby/2.0.0/universal-darwin15", "/System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/lib/ruby/vendor_ruby", "/System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/lib/ruby/2.0.0", "/System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/lib/ruby/2.0.0/x86_64-darwin15", "/System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/lib/ruby/2.0.0/universal-darwin15"]
irb(main):002:0> require 'global'
TypeError: no implicit conversion of nil into String
        from /usr/local/Library/Homebrew/config.rb:34:in `initialize'
        from /usr/local/Library/Homebrew/config.rb:34:in `new'
        from /usr/local/Library/Homebrew/config.rb:34:in `<top (required)>'
        from /System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/lib/ruby/2.0.0/rubygems/core_ext/kernel_require.rb:55:in `require'
        from /System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/lib/ruby/2.0.0/rubygems/core_ext/kernel_require.rb:55:in `require'
        from /usr/local/Library/Homebrew/global.rb:18:in `<top (required)>'
        from /System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/lib/ruby/2.0.0/rubygems/core_ext/kernel_require.rb:55:in `require'
        from /System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/lib/ruby/2.0.0/rubygems/core_ext/kernel_require.rb:55:in `require'
        from (irb):2
        from /usr/bin/irb:12:in `<main>'
```

**Comment by jeroenj** [comment]
Reading a little further the Chef cookbook's issue is a bit different since it does not use `bin/brew` which sets the desired environment variables.

**Comment by mikemcquaid** [comment]
@jeroenj That's a really bad idea of the Chef cookbook. Can you try `cd /usr/local && git reset --hard origin/master` and then see if that fixes it when you run `brew --prefix`?

**Comment by jeroenj** [comment]
For me all `brew` commands were working correctly. I was bein a bit too optimistic when posting the comment about the chef cookbook. I think it has nothing to do with @tmongero's issue. I'm going through the source of the cookbook right now. It looks like it's quite heavily using the homebre internals. :-/

**Comment by tmongero** [comment]
@uniqMartin There was an error in an update I tried to run just before this all began, but unfortunately I don't have it anymore as it was pre-upgrade (& thus, reboot). I do remember doing `brew update` and getting an error explaining that it could not update, then I did brew doctor and it mentioned something about not having access to /usr/local/ - I did `sudo chown -R myusername /user/local/` and that didn't help.

I attempted to run `cd /usr/Local && git reset --hard origin/master` and received the following output:

![image](https://cloud.githubusercontent.com/assets/16004728/12432367/7480cc0a-bec0-11e5-8373-e95e074cbd67.png)

**Comment by djthread** [comment]
I had the same issue as OP and was able to resolve it with git reset --hard origin/master.

Thanks to tdsmith from #machomebrew for pointing this issue page out to me.

**Comment by UniqMartin** [comment]
@tmongero Yes, that's a nasty permission error. Please don't run the commands I suggested with `sudo`. This will only create more permission problems, unfortunately. The following command sequence should in theory solve the issue (there was a typo in your comment):

```
sudo chown -R $(whoami) /usr/local
cd /usr/local && git reset --hard origin/master
```

Make sure the first one completes without error and preferably check if it had the desired effect by looking at the output of, e.g., `ls -AlF /usr/local`.

**Comment by tmongero** [comment]
You solved it, @UniqMartin !! That worked - Thanks so much!!

**Comment by UniqMartin** [comment]
@tmongero Good to hear you were able to resolve the issue now! :smiley_cat:

**Comment by deinspanjer** [comment]
I also just ran into this issue "out of the blue", and I had to fix the permissions on /usr/local then do a git reset to fix it.

I wanted to share one additional point in case it helps track down the cause.  I normally run brew commands by first starting a normal Terminal window.  In this particular case, I had just updated Alfred 2 and while browsing through the preferences, I noticed it had a set of workflows for Homebrew.  I tried running the "brew update" workflow from within Alfred.  I'm not sure if it was something built in to Alfred or if it was maybe added by Brew Cask.

After running that brew update command inside Alfred 2, it had an error, and then my entire brew was non-functional until I fixed it with the above commands.
